pkg/controllers/statusaggregator/plugins/pod.go

```diff
@@ -0,0 +1,153 @@
+) (*unstructured.Unstructured, bool, error) {
+	logger := klog.FromContext(ctx).WithValues("status-aggregator-plugin", "pods")
+
+	needUpdate := false
+	aggregatedStatus := &corev1.PodStatus{}
+	phases := map[corev1.PodPhase][]string{
+		corev1.PodPending:   make([]string, 0, len(clusterObjs)),
+		corev1.PodRunning:   make([]string, 0, len(clusterObjs)),
+		corev1.PodSucceeded: make([]string, 0, len(clusterObjs)),
+		corev1.PodFailed:    make([]string, 0, len(clusterObjs)),
+		// Ignore phase PodUnknown as it has been deprecated since year 2015.
+	}
+
+	for clusterName, clusterObj := range clusterObjs {
+		logger := klog.FromContext(ctx).WithValues("cluster-name", clusterName)
+
+		utd := clusterObj.(*unstructured.Unstructured)
+		// For status of pod
+		var found bool
+		status, found, err := unstructured.NestedMap(utd.Object, common.StatusField)
+		if err != nil || !found {
+			logger.Error(err, "Failed to get status of cluster object")
+			return nil, false, err
+		}
+
+		if status == nil {
+			continue
+		}
+
+		podStatus := &corev1.PodStatus{}
+		if err = runtime.DefaultUnstructuredConverter.FromUnstructured(status, podStatus); err != nil {
+			logger.Error(err, "Failed to convert the status of cluster object")
+			return nil, false, err
+		}
+
+		if podStatus.Phase == "" {
+			// If the status has not yet been generated, set it to pending as default.
+			phases[corev1.PodPending] = append(phases[corev1.PodPending], clusterName)
+			continue
+		}
+
+		phases[podStatus.Phase] = append(phases[podStatus.Phase], clusterName)
+		if aggregatedStatus.StartTime.IsZero() || podStatus.StartTime.Before(aggregatedStatus.StartTime) {
+			aggregatedStatus.StartTime = podStatus.StartTime
+		}
+		for _, initContainerStatus := range podStatus.InitContainerStatuses {
+			initContainerStatus.Name = fmt.Sprintf("%s (%s)", initContainerStatus.Name, clusterName)
+			aggregatedStatus.InitContainerStatuses = append(aggregatedStatus.InitContainerStatuses, initContainerStatus)
+		}
+		for _, containerStatus := range podStatus.ContainerStatuses {
+			containerStatus.Name = fmt.Sprintf("%s (%s)", containerStatus.Name, clusterName)
+			aggregatedStatus.ContainerStatuses = append(aggregatedStatus.ContainerStatuses, containerStatus)
+		}
+	}
+
+	// Check phase in order: PodFailed-->PodPending-->PodRunning-->PodSucceeded.
+	for _, phase := range []corev1.PodPhase{corev1.PodFailed, corev1.PodPending, corev1.PodRunning, corev1.PodSucceeded} {
+		if len(phases[phase]) == 0 {
+			continue
+		}
+		if aggregatedStatus.Phase == "" {
+			aggregatedStatus.Phase = phase
+		}
+		sort.Strings(phases[phase])
+		if aggregatedStatus.Message == "" {
+			aggregatedStatus.Message = fmt.Sprintf("Pod %s in clusters [%s]",
+				strings.ToLower(string(phase)), strings.Join(phases[phase], ","))
+			continue
+		}
+		aggregatedStatus.Message = fmt.Sprintf("%s, and %s in clusters [%s]",
+			aggregatedStatus.Message, strings.ToLower(string(phase)), strings.Join(phases[phase], ","))
```

Comment:
are we missing a `break` here?

Reply:
oh i misinterpreted... i thought that we were replacing the  phase and message with each loop. this is actually ok